### DISCUSSION TITLE
Add Control Change to MIDI Nodes

### DISCUSSION THREAD

**Discussion by jamesa08** [JacquesLucke/animation_nodes#1683] [Ideas]
I have been doing a lot of MIDI work in Animation Nodes and have found myself using separate MIDI tracks to trigger events in the scene. I also am finding the lack of modulation and pitch bend frustrating. I believe these events can be implemented with Control Change. It is documented on the MIDI Association website, found here: https://www.midi.org/specifications-old/item/table-3-control-change-messages-data-bytes-2.

I am aware that pitch bend is not in the MIDI CC standard, perhaps this can be implemented somehow?

Thanks!

**Comment by OmarEmaraDev** [comment]
How should we expose such information to the user? Can you suggest a design that incorporate such information?

**Comment by jamesa08** [comment]
Sure. Heres what I made in Photoshop. I was thinking either using a style of brackets to determine the Control Change (such as ['modulation']) just like to set other parameters. Another way I was thinking was just using the control change number, which could prove to be confusing. Or the last method (and how my DAW does it) use a list to select the Midi CC. 
Hope this helps! 
<img width="672" alt="ControlChange" src="https://user-images.githubusercontent.com/79052050/111053976-96522b80-8436-11eb-956a-379d30091daf.png">

**Comment by OmarEmaraDev** [comment]
Can you provide a description for a practical example for those features?

**Comment by jamesa08** [comment]
Heres one of the animations I've been working on for the past 2 months. Currently, I'm using a separate MIDI track to trigger when the vibraphone arms should come up/down. This could easily just be put in the original MIDI track using MIDI CCs.
https://user-images.githubusercontent.com/79052050/111075825-24262900-84c0-11eb-85a8-7b8d0eba140d.mp4

Another example would be vibrating a laser back and forth when the modulation wheel is present or gliding the notes with the pitch wheel (which would change the rotation of the laser).

Others include triggering objects in the scene to do something (such as lights turning off / on, triggering animations, etc). 
Having MIDI Control Change would be very beneficial to the current set of MIDI Nodes. 

I have also realized in my design that Control Change is not on a per note basis, it is on a per track basis. I have updated what a MIDI Track Node could look like instead of having its own node. 
<img width="617" alt="ControlChange2" src="https://user-images.githubusercontent.com/79052050/111076552-20e06c80-84c3-11eb-9acf-be93bc98ede9.png">

**Reply by Majoran99** [reply]
Hi, cool video - can I ask you, how do you achieve to create again and again new ball objects triggered by the MIDI data? (and terminating them when out of use)

**Reply by jamesa08** [reply]
I will release a video on my Youtube channel when I finish my project. I’ll link you the video when its out!

**Reply by Majoran99** [reply]
cool! :-)

**Comment by OmarEmaraDev** [comment]
@imacj Sounds good, but the newly suggested design is not correct either. CCs happen per channel of a track, not globally in a track. Moreover, If multiple Controller events happen on the same channel, then the CCs will change across time, and not uniformly. So a single output CC value in the Track Info node wouldn't work.
